lib/simulator_sse.h

```diff
@@ -0,0 +1,599 @@
+  // The inner loop (V_i = \sum_j M_ij V_j) is unrolled by hand.
+  // Performs full SSE vectorization.
+  void ApplyGate1H(unsigned q0, const fp_type* matrix, State& state) const {
+    uint64_t sizei = uint64_t{1} << num_qubits_;
+    uint64_t sizek = uint64_t{1} << (q0 + 1);
+
+    uint64_t mask0 = sizek - 1;
+    uint64_t mask1 = (2 * sizei - 1) ^ (2 * sizek - 1);
+
+    fp_type* rstate = StateSpace::RawData(state);
+
+    auto f = [](unsigned n, unsigned m, uint64_t i,
+                uint64_t sizek, uint64_t mask0, uint64_t mask1,
+                const fp_type* matrix, fp_type* rstate) {
+      uint64_t si = (16 * i & mask1) | (8 * i & mask0);
+
+      __m128 r0, i0, r1, i1, ru, iu, rn, in;
+
+      uint64_t p = si;
+      r0 = _mm_load_ps(rstate + p);
+      i0 = _mm_load_ps(rstate + p + 4);
+      ru = _mm_set1_ps(matrix[0]);
+      iu = _mm_set1_ps(matrix[1]);
+      rn = _mm_mul_ps(r0, ru);
+      in = _mm_mul_ps(r0, iu);
+      rn = _mm_sub_ps(rn, _mm_mul_ps(i0, iu));
+      in = _mm_add_ps(in, _mm_mul_ps(i0, ru));
+      p = si | sizek;
+      r1 = _mm_load_ps(rstate + p);
+      i1 = _mm_load_ps(rstate + p + 4);
+      ru = _mm_set1_ps(matrix[2]);
+      iu = _mm_set1_ps(matrix[3]);
+      rn = _mm_add_ps(rn, _mm_mul_ps(r1, ru));
+      in = _mm_add_ps(in, _mm_mul_ps(r1, iu));
+      rn = _mm_sub_ps(rn, _mm_mul_ps(i1, iu));
+      in = _mm_add_ps(in, _mm_mul_ps(i1, ru));
+      p = si;
+      _mm_store_ps(rstate + p, rn);
+      _mm_store_ps(rstate + p + 4, in);
+
+      ru = _mm_set1_ps(matrix[4]);
+      iu = _mm_set1_ps(matrix[5]);
+      rn = _mm_mul_ps(r0, ru);
+      in = _mm_mul_ps(r0, iu);
+      rn = _mm_sub_ps(rn, _mm_mul_ps(i0, iu));
+      in = _mm_add_ps(in, _mm_mul_ps(i0, ru));
+      ru = _mm_set1_ps(matrix[6]);
+      iu = _mm_set1_ps(matrix[7]);
+      rn = _mm_add_ps(rn, _mm_mul_ps(r1, ru));
+      in = _mm_add_ps(in, _mm_mul_ps(r1, iu));
+      rn = _mm_sub_ps(rn, _mm_mul_ps(i1, iu));
+      in = _mm_add_ps(in, _mm_mul_ps(i1, ru));
+      p = si | sizek;
+      _mm_store_ps(rstate + p, rn);
+      _mm_store_ps(rstate + p + 4, in);
+    };
+
+    ParallelFor::Run(num_threads_, sizei / 8, f,
+                     sizek, mask0, mask1, matrix, rstate);
+  }
+
+  // Applies a single-qubit gate for qubit <= 1.
+  // Performs vectorized sparse matrix-vector multiplication.
+  // The inner loop (V_i = \sum_j M_ij V_j) is unrolled by hand.
+  // Performs partial SSE vectorization with permutations.
+  void ApplyGate1L(unsigned q0, const fp_type* matrix, State& state) const {
+    uint64_t sizei = uint64_t{1} << (num_qubits_ + 1);
+
+    fp_type* rstate = StateSpace::RawData(state);
+
+    auto f = [](unsigned n, unsigned m, uint64_t i, unsigned q0,
+                const fp_type* matrix, fp_type* rstate) {
+      __m128 r0, i0, r1, i1, ru, iu, rn, in, rm, im;
+
+      uint64_t p = 8 * i;
+
+      r0 = _mm_load_ps(rstate + p);
+      i0 = _mm_load_ps(rstate + p + 4);
+
+      switch (q0) {
```

Comment:
This `switch` appears to omit several cases - could you add an empty default case or a comment to make it clear that this is intentional?

Reply:
The same applies to other `switch` statements in this file.

Reply:
The `switch` statements do not omit any cases. I added empty default cases.